### ISSUE TITLE
uMatrix icon only available in one browser window when auto-restoring multi-window session

### ISSUE THREAD

**Issue by DjogaRo** [gorhill/uMatrix#357]
Hi there,
first, thanks for this add-on! When I learned about it, I even considered using Firefox again - well for a moment at least. Love that you made the efford to make it compatible with Pale Moon.

I'm using Pale Moon 25.7.0 (x64) with TabMixPlus on Win7 Pro SP1 64-Bit.
*Edit*: uMatrix 0.9.2.1
TMP might be relevant, since I use its session manager. *Edit/Update: It is irrelevant.*

The problem manifests, when I start Pale Moon and a previous session containing multiple windows gets automatically restored. Only the first window gets to have the uMatrix icon. For the other windows not having the icon means, it's neither where I placed it (if I placed it) nor does it show up in the "Customize Toolbar" icon pool window.

When I open additional windows, those will have the icon. When I restored a 3-window session manually, 2 out of 3 windows did have the icon.

I use my setup for years now, my ongoing session having more then one window ain't anything new either, and I use several add-ons that come with buttons I placed in the toolbar, but for this behaviour I cannot give another example.

I hope there's a straight forward fix to this and coming soon, since TabMixPlus restores a sessions windows in weird order and I seldomly get the icon in the most important one.

**Update**:
I just disabled TabMixPlus, closed all tabs and windows, restarted, set Pale Moon to restore last time tabs and windows on start, opened some windows with a few tabs (at that point all windows had the icon), exited PM, started it again, and experienced the same problem.
So TabMixPlus seems not to be relevant.

After that I re-enabled TMP, restarted Pale Moon - only one window with pinned tabs at that moment -, restored my main session manually (not replacing the open window). I noticed that now it was reversed which window had the icon and which not. The already open window did have it, the first from the session that got opened did not, the second and third did, curious.

**Update**:
*Sidenote: uBlock Origin suffers the same problem, plus being a bit forgetfull regarding its settings. I've got a feeling the n/a-thing and the forgetfullness might share some common ground. The latter I will make an issue in the other project.*

Pale Moons first start after booting Windows today was different. The last session (3 windows) got auto-restored as usual, but the problem didn't manifest at all. All 3 windows had the icons of uMatrix and uBlock_o on the addressbars right side, where I put them. *Also uBlock_o had re-remembered its settings.* I promptly exited PM and started over and both problems were back. That's why I tend to assume common ground.
BTW, what I've reported before this 2nd update all happened with uBlock_o not being installed yet.

By now I'd say when PM starts for the 2nd, 3rd etc. time during a Windows session not all events get fired you rely upon for initialization. ? Or you register differently than other add-ons do? (just guessing; don't know the API)

Next I'll try with fresh, clean PM profiles, try it from a problem-free state closing a window and restoring it through History -> Recently closed windows ..., and maybe I'll install FF and try to reproduce the problem there. Till then ...

**Comment by DjogaRo** [comment]
**Testing yielded two significant aspects: Pinned Tabs and restoring a window through History**

Did some testing with a new Pale Moon profile and uMatrix only.
I didn't change uMatrix' default settings.
I changed PMs Home Page to "google.com/ncr", which resolves to "https://www.google.com/?gws_rd=ssl". I only used this page in every tab during testing.
And I changed the PM start up setting to restore windows and tabs from last time.

Having 2 windows with 1 tab each, exiting and starting PM again yielded no problem, i.e. both windows kept the uMatrix icon.
But the problem can be reproduced having 2 tabs in the first window with the first one being **pinned**. *(Having 2 tabs in first window, none pinned -> no problem)*

Then I tried to reproduce the problem when closing a window and restoring it through History -> Recently Closed Windows ...
Strangely enough that procedure has the opposite effect. It doesn't produce the problem, it repairs it. So if I yielded the problem through exit PM, start PM, and then closed the window with the icon missing to restore it through history, the restored window got the icon. That "healing effect" even lasts through the next exit-start-PM iteration. So the latter has to be done 2 times to have a problem again.

Next thing on my list: doing the same with FF. Till then ...

**Comment by DjogaRo** [comment]
Tried the same testing with **Firefox 40.0.3**.
**Could not reproduce.**

**Comment by DjogaRo** [comment]
Same test, **Firefox 24.8.1esr**
**Reproducible.**
Basically the same as with Pale Moon, but no need for pinned tabs, thus easier to reproduce. Although, with FF the problem was a bit fluctuating.
First it was that after restoring a problematic window through History -> Recently Closed Windows ... it often took 3 instead of 2 iterations of exiting and starting FF to have an icon missing again. Then although uMatrix' default settings allow 1p cookies, when starting FF the restored tabs got redirected to google.de. And after that episode the history restore had lost its lasting effect, i.e. now it constantly took only one exit-start-FF to have a window missing the icon.

All very strange.

**Comment by DjogaRo** [comment]
For a few starts of PM now my problems with uMatrix and uBlock0 seem to be gone (including the mentioned forgetfulness). My current version of PM is 25.7.2 (x64), but i can't tell if the problems' end coincided with a PM-update. uMatrix still shows version number 0.9.2.1.

Though those problems don't bother me anymore, I confirmed that my last post is still valid. Well, except for the PM reference.

**Comment by DjogaRo** [comment]
Uhm, that might have been a bit premature. PM just got updated to 25.7.3 and I got both problems back.

**Comment by gorhill** [comment]
I did some work re. toolbar button in uBlock Origin ([1.3.0rc4](https://github.com/gorhill/uBlock/releases/tag/1.3.0rc4)). If this resolves the toolbar button issues for uBlock Origin, I will just import these fixes in uMatrix.

**Comment by DjogaRo** [comment]
Thanks for your response.
So I just made a new profile and installed uBlock0 from the link you provided and went through my test routine.
Whatever you did definetly has an effect. Now when I cycle through (Main Menu -> "Exit", Start PM) the toolbar button is missing every second time. When I mess around like changing which window's Main Menu I use to exit, it sometimes needs a few iterations to fall back into this pattern. But apart from that the pattern is stable. So, not statistically every second time, but literally.
Strange, but a change. :)

**Comment by DjogaRo** [comment]
Same test, Firefox 24.8.1esr
Same same, but different. Here it's that strict, alternating pattern. It seems to be a random series of the icon being there, not being there.

**Comment by DjogaRo** [comment]
**Correction**:
Further playing around yielded the missing of the button in all but one window on auto-session-restore is random in PM, too.
That also goes for the test profiles with uMatrix 0.9.2.1 or uBlock0 1.2.1 respectively. So I guess it has always been random for me and presented a pattern by chance or some things do have kind of an influence or changes to PM might have caused a pattern shift. 

However, you indirectly led me to https://github.com/gorhill/uBlock/issues/833. And, you saying
> I added another chunk of code to further defer adding the toolbar icon at a later time if uBlock can't find the required component when a new window is created.
made something else catching my eye.

old := in PM with uBlock0 1.2.1
new := in PM with uBlock0 1.3.0rc4

The old behaviour when starting PM, in every restored window - in which the button gets to be - said button starts out being empty bearing no picture yet and only having little extent, as if maybe caused by some set padding. Then the picture appears; the button is complete.
The new behaviour is, the button starts out complete and then continue with the old behaviour from its beginning, i.e. being empty for a fracture of a second.

And that's actually all the difference I can make out so far.
On opening new windows I cannot make out such thing happening (maybe just faster?). And I don't see it happening this way in FF24.

I also read the PM-Forum thread linked in uBlock issue 833. Others report to have the button missing in their only window and in new windows, which never was the case for me. The first restored window on PM start always has it. So does every new window I open.

In my standard profile with my standard session I seemingly always get the button in only the first window. But that's loaded with some more extensions plus the first window has 83 tabs - 27 of them pinned - and the second has another 20 tabs. Some of that might shift the odds.

**Comment by DjogaRo** [comment]
**Update**:
The button placing code seems **not** to be a problem here. After I found out, that one can use unpacked add-ons and got the browser to show me some logging I put in, I observe that addLegacyToolbarButton(...) gets called as many times as I get windows bearing the button, i.e.
**Whenever I got the button missing in a window, it is because addLegacyToolbarButton(...) hasn't been executed for that window.**

Is that the point where I change this issue's title? If so, to what?

**Comment by gorhill** [comment]
> Whenever I got the button missing in a window, it is because addLegacyToolbarButton(...) hasn't been executed for that window.

Thanks, that helps of narrow down where to look in the code. I will publish an interim build later just for the sake of checking whether the problem is fixed with further changes.

**Comment by gorhill** [comment]
[rc6 on _Releases_ page](https://github.com/gorhill/uBlock/releases/tag/1.3.0rc6) -- will check feedback later when I am back.

**Comment by DjogaRo** [comment]
Thanks. Did the testing and unfortunately could still reproduce.
I'll pave that thing with console.log() statements and poke around later this day.

I haven't mentioned this here, yet, since I was going to make it a separat issue, but in the new context...
When uBlock0 is missing its button, it also misses to use "My Filters". When I open the dashboard they are listed, but only after adding sth. and applying changes, the list takes effect again.

**Comment by DjogaRo** [comment]
Right now I'm at work, but for later it would be great to know the entry points already - for when their is a new browser instance and for when their is a new window. So, if you can provide them from the top of your head, it would help me a lot.

**Comment by gorhill** [comment]
Just to be sure... Are you testing using uBlock?

This is the entry point for when a new chrome window is created:
https://github.com/gorhill/uBlock/blob/master/platform/firefox/vapi-background.js#L1159

**Comment by DjogaRo** [comment]
Yes|No, I'm testing using uBlock Origin to be exact. :)
Both Buttons (uBlock0, uMatrix) are always present together or missing together. Well since there's an exception to everything, This morning in my second window (of two) uBlock's was there, uMatrix' Button was not. So, I presume, it doesn't matter which of them I use, does it.?

And thanks for the info.
(still not back home)

**Comment by gorhill** [comment]
As said all the attempted fixes are in uBO, not in uMatrix, nothing has changed in uMatrix. If the fixes work for uBO, then I will import them to uMatrix. The point is to validate whether the fixes in uBO works in latest release.

**Comment by DjogaRo** [comment]
In case of a missing button onWindowLoad() did not get executed for that window.

**Comment by gorhill** [comment]
You _console.logged_ uBO, right?

If so, the only reason that I can think for this to happen is that Firefox is not passing the `domwindowopened` event to uBO. This event is [received and handled at line 1219](https://github.com/gorhill/uBlock/blob/master/platform/firefox/vapi-background.js#L1219), in `windowWatcher.observe`. I don't see how this could be the case, hence I did not think it was useful to _console.log_ this. Maybe you could try to just confirm the event is indeed really not fired, or if it is fired, see what is the value of `topic`?

**Comment by DjogaRo** [comment]
Correct, uBO 1.3.0.

vAPI.tabs.getWindows producec exactly one window, that would be why the first to be restored window always gets the button. But windowWatcher.observe does not always get executed for the remaining ones. (I put the logging at this functions very beginning.)

Would be interesting to see what happens, when first the windowWatcher gets registered and then the intial loop over existing windows gets done. I'll try that tomorrow. Just to rule out, that the remaining windows are produced between said loop and the registration.

**Comment by DjogaRo** [comment]
Just tried the latter. Makes no difference.

**Comment by gorhill** [comment]
> vAPI.tabs.getWindows produces exactly one window

Hmm interesting. I see a test against a `closed` attribute [at line 691](https://github.com/gorhill/uBlock/blob/master/platform/firefox/vapi-background.js#L691), I wonder if this could be related...

**Comment by DjogaRo** [comment]
I made it test for (true || !win.closed), but still only one windows produced. In cases of error and cases of success (regarding button's presence). I really gotta get some sleep. Til then.

**Comment by gorhill** [comment]
> I really gotta get some sleep. Til then.

Thanks for taking the time to investigate. When people ask me what they could contribute, this is pretty much it, investigating issues -- especially the ones I can't reproduce.

So at this point maybe a Pale Moon dev could provide some insights as to why `getEnumerator('navigator:browser')` returns only one window while no `domwindowopened` is not fired for the other ones.

**Comment by DjogaRo** [comment]
Not for that. That add-on itself is a huge chunk of work. What's investigating a problem, that bugs me, compared to building that thing in the first place. :) As soon as I knew, how I could investigate, there was no reason not to.

Yeah, I'm going to make a thread at PM's forums later. Maybe the API will get an update or there might be a well trodden way to handle the given behavior. We'll see.

**Comment by DjogaRo** [comment]
I was just about to report in the PM forum, when I recognized, that getting the initiyl windows utilizes an other service then the registering an observer. Unification fixes the problem. Now my second window sometimes gets handled in the initial loop in start() and sometimes through an event by the observer. I'd yet have to see it handled in both codepathes during one startup.

Just replace
```javascript
vAPI.tabs.getWindows = function() {
    var winumerator = Services.wm.getEnumerator(this.chromeWindowType);
    var windows = [];
```
with
```javascript
vAPI.tabs.getWindows = function() {
    var winumerator = Services.ww.getWindowEnumerator(this.chromeWindowType);
    var windows = [];
```

In an example utilizing wm a listener for the windows load event was registered with the window whenever there's a new window opened. I have a feeling, that way all those delays may be unneeded. I'll investigate.
The example is the chosen answer in:
http://stackoverflow.com/questions/29178158/in-a-firefox-restartless-add-on-how-do-i-run-code-when-a-new-window-opens-list

At PM forum I also asked, whether WindowWatcher or WindowMediator was to be preferred - respecting that the code is meant for all supported FF-lings. The thread is:
http://forum.palemoon.org/viewtopic.php?f=16&t=9893
(I also asked to move the thread, and the way the url looks it may change by that).

**Comment by gorhill** [comment]
> I recognized, that getting the initial windows utilizes another service than the registering an observer

Yes, I saw this yesterday, and wondered too if this could be the cause. Maybe there was a good reason to have been done this way, I will have to go through the commit history to see if this was like this since the beginning.

Once I release 1.3.1 I will revise teh code and thsi will have to be well tested to find out if there are any side effect, including for Firefox for Android.

**Comment by DjogaRo** [comment]
Let me guess. Occasionally having two entries in the second window's context menu?
When the second window gets handled by the initial loop in start and there's a delay because `( wintype !== 'navigator:browser' )` evaluates to true, it somehow happens, that it gets two CM entries.

I have yet to understand the magic behind those entries. With the same delays the first window still ends up with one entry. If I comment out the entry's registration in attachToTabBrowser, the first window still gets it - even the second window does sometimes.

Searching in files for "contextMenu.register" finds me no other call.

**Comment by gorhill** [comment]
Alright, I will push a new revision for uBO, now strictly using WindowMediator. I tested from Palemoon to Nightly, legacy toolbar button/Australis, and all seems to work just like before. Hopefully though, this will fix the issue on your side.

The doubling of the menu item is easily taken care of by first checking whether the item is already inserted.

**Comment by gorhill** [comment]
Can you confirm whether the fix work using [latest uBO dev build](https://github.com/gorhill/uBlock/releases/tag/1.3.2b1)?

**Comment by DjogaRo** [comment]
It's not here anymore. I'v got to wait till for the FF-release of 1.3.2b2, I guess.

**Comment by gorhill** [comment]
1.3.2b2 is available now in the _Releases_ section.

**Comment by DjogaRo** [comment]
I can't see an xpi there. so I copied 1.3.0 to a new profile and replaced vapi-background.js with the on from 1.3.2b2.
With that the issue still occurs from time to time.
As soon as I get my hands on the new xpi or you tell me the mix I did suffices, I'll look into it.

**Comment by DjogaRo** [comment]
I'm on it

**Comment by gorhill** [comment]
> I can't see an xpi there

My bad, I failed to notice the build failed for some reasons.

**Comment by DjogaRo** [comment]
Nevermind.
The second window sometimes is neither handled in the initial loop nor does onOpenWindow get called.
And still it gets a context menu entry. That's way above me.

Is WindowMediator to be prefered over WindowWatcher? Since using the other one, initial loop and event handling complemented each other quite good.

**Comment by gorhill** [comment]
I suppose both could be used, but I would have to review the code in there to be sure the it's proofed against multiple requests to attach to a window -- i.e. check if whatever needs to be done is already done.

**Comment by gorhill** [comment]
> And still it gets a context menu entry

Yes, because uBO's core code calls for the installation of the context menu entry explicitly.

**Comment by DjogaRo** [comment]
Well, I had already tried (with WW) making the observer add an event listener to the new window - which is to wait for a "load" event. The listener function I had defined within observe(...), which I couldn't figure out to do in start(), too. That way onWindowLoad(...) when called by the observer never showed delays happening in the attach... functions.

So now I added
```javascript
	var getOnLoadListener = function(win) {
		function onLoadListener() {
			console.log("__LL: Got called for a window done loading!");
			win.removeEventListener("load",onLoadListener);
			// Only add UI changes if this is a browser window
			if (win.document.documentElement.getAttribute("windowtype") === "navigator:browser") {
				console.log("__LL: Event domwindowopened");
				onWindowLoad(win);
			} else
				console.log("__LL: windowtype ain't navigator:browser!");
		}
		return onLoadListener;
	}
```
and replaced `onWindowLoad(win);` with `win.addEventListener("load", getOnLoadListener(win));` so that I can use that in start, too. Having two context menu entries in window 2 only happened when handled in start and later experiencing delays. This should fix it plus make the delay code obsolete.

But now. I never see the initial loop iterated twice. Instead, I get
```
[20:31:40.211] this._cps2 is undefined @ chrome://browser/content/browser.js:1614
[20:31:40.405] this._lastStatus is null @ chrome://browser/content/browser.js:14018
[20:31:40.406] this._cps2 is undefined @ chrome://browser/content/browser.js:1614
[20:31:40.407] this._lastStatus is null @ chrome://browser/content/browser.js:14018
```
or one handled through the loop and one through the observer and then all goes fine.
I don't understand how these errors in browser.js come about, but I've got a feeling, I have to look for them using the unchanged 1.3.2b2 again.

**Comment by gorhill** [comment]
> This should fix it plus make the delay code obsolete.

The delay code has to stay. I just want to know if the latest b3 solves the issue.

**Comment by DjogaRo** [comment]
I'll have a look.
By the way, I've reverted 1.3.2b2 back to before my additions. Those errors in browser.js always happen - when buttonwise all is fine and when not.

**Comment by DjogaRo** [comment]
Still missing the button in the 2nd window sometimes.
Looking at the code changes I don't get which issue you were addressing.

**Comment by gorhill** [comment]
> Looking at the code changes I don't get which issue you were addressing.

I think I am getting lost in the info here. Isn't true that not all windows are reported in the event handler? I thought you were saying here that both windowWatcher and windowMediator together would result in all windows being reported to `onWindowLoad()`.

**Comment by DjogaRo** [comment]
I'm get lost, too. :)

No, mixing those two doesn't work here, cause using
```javascript
vAPI.tabs.getWindows = function() {
    var winumerator = Services.wm.getEnumerator(this.chromeWindowType);
```
for startup never yields more than one window, for whatever reason. Using WindowWatcher for that, too, gives also those windows during startup, which won't cause a notification to a registered observer later. So only using WW, restored session windows are early enough generated to get them initially or certainly handled later.

In 1.3.2b2 you tried, if both places utilizing WM would also fit together, but it didn't. At least not on my machine.

And me getting lost shows in that I had forgotten to replace wm when getting the enumerator. Fixed this. Now all is fine (using onLoadListener), except for that sometimes now the first window gets no button. So I managed to invert the problem. :)

And I noticed there are some more places utilizing either Services.wm or Services.ww. And I have no idea, where this mixture doesn't matter and where it might.

**Comment by DjogaRo** [comment]
Regarding my own attempt, I suspect, that the first window sometimes already fired the "load" event, when I'm registering for that. At least that would explain my outcome.
So now I'm trying to find out how to test for the event having already been fired to call onWondowLoad directly. Or I could do the latter everytime in the loop's first iteration.

Eitherway, I'll report back. And btw, I didn't throw out the delay code, I keep that spiked with logging. In the end I want to know, if what I try is usefull.

**Comment by DjogaRo** [comment]
It seems to work fine. I haven't tested excessively, but so far I only saw exactly one context menu entry per window. Same goes for the toolbar button. And the log shows no going down the delay codepaths.
If you want to take a look:
https://gist.github.com/DjogaRo/15b6527ecdd6da278ec6

Testing on legacy FF I'll do at a later time.

I'm off to see Morpheus now.

**Comment by gorhill** [comment]
I already rewrote the window watching code. Will push a new build tomorrow.

**Comment by gorhill** [comment]
In latest build, I use both winddowMediator and windowWatcher's enumerators and event handlers, so there should be no cracks left for the events to fall through.

**Comment by DjogaRo** [comment]
I think I'll be home in one or two hours. I'll test it then.

**Comment by DjogaRo** [comment]
There again were instances of the missing button. So I wanted to come back already knowing, where it comes from.
I again added logging and a minor change to test my suspect - here's the diff (basis was uBO 1.3.2b5):
https://gist.github.com/DjogaRo/c5764a5469c8c4be54be
and a would have been problematic PM startup's logging output (the wording is not to be taken too seriously):
https://gist.github.com/DjogaRo/b99b2ee995727d8ee176
(the 3rd window is the browser console)

With the tests you apply to the results, the window watcher's enumerator gives you, you skip the window you are after. It's to early for the window to satisfy those tests - at least on my machine.

Using it exactly like the other one (enumerators) should do the trick. And why not `Services.ww.getWindowEnumerator(chromeWindowType)`?

I guess it doesn't matter that early, but with the 2nd enumerator you change the ids of the windows you've already got.

**Comment by gorhill** [comment]
> It's to early for the window to satisfy those tests - at least on my machine

Argh, that's right, I actually really meant to not do that test, given that the handler will delay until the test is satisfied -- or give up. Fo0 got to remove it this morning.

> why not `Services.ww.getWindowEnumerator(chromeWindowType)`?

Because the [doc](https://developer.mozilla.org/en-US/docs/Mozilla/Tech/XPCOM/Reference/Interface/nsIWindowWatcher#getWindowEnumerator%28%29) says there are no parameters:

> Parameters
> None.

**Comment by gorhill** [comment]
Remove the test in latest b6 release.

**Comment by DjogaRo** [comment]
I just made lots of exit-start with a new profile, my standard test scenario, and b6. As to be expected by now all went fine, seemingly no issue left. I did no logging with b6, but since what I tried in b5 and removing those evil test is equivelent, I'd say it's done.

Thanks so much for seeing this through. It takes away an annoying inconvenience for me that came with almost every browser startup bearing my bloated standard session. :)

On a sidenote still, out of curiosity: Why not getting rid of those delays? Are there supported hosts, that don't fire the "load" event for windows?

**Comment by gorhill** [comment]
> Why not getting rid of those delays?

Because it's more foolproof than cascading on yet another specific event. The code used to wait for `DOMContentLoad`, which [could also delay](https://github.com/AlexVallat/uBlock/commit/5ac4ec21e9e960346b72889e842dbb6b0772aa76?diff=unified#diff-8334c71c60f201b518259aad4d0bddd8R1527) (once) in case some conditions were not fulfilled. Given that uBO is meant to run on so many different flavors of FF, the time-based event handling will work on all platforms, it's the most generic and foolproof one: try until the environment is sane (or give up after it fails too often). The toolbar button is not critical for the good functioning of uBO, it can be handled asynchronously, and delayed if needed.

**Comment by gorhill** [comment]
Ok now I will have to import all this in uMatrix. Not sure when I will have time to do this.

**Comment by DjogaRo** [comment]
Alright. But I know it's coming and that's great.

Regarding the delays, I'm not bothered with it's time aspect at all. I just think having one event handler for new windows registering a second event handler for adding what needs to be added is a much cleaner solution coding-wise. Given there is a proper event flying around. And I myself am pretty convinced the window's "load" event is damn appropriate. (It's not equivalent to DomContentLoaded, and not meant to be.)
And with code cleanliness I'm bothered, because it helps with maintainability. Seeing you rotating taking care of issues, progressing the wiki and whatnot, I figure you need every bit of maintainability you can get. What I'm actually worried about is your ongoing well-being. :)

**Comment by gorhill** [comment]
> I just think having one event handler for new windows registering a second event handler for adding what needs to be added is a much cleaner solution coding-wise.

I think the opposite -- it's currently cleaner. Given that the need for delaying is inevitable anyways, why bother dealing at all with an intermediate event which is rendered useless by the need to delay?

**Comment by DjogaRo** [comment]
*Well, it's your baby, so I don't expect you to do anything at all and I understand you're already busy as hell with your projects. But if you feel intrigued and find the time to further reflect on this, I'll have elaborated :)* :

That's a matter of perspective or paradigm, I guess. To me the first event handler has it's own proper job to do regarding new windows that pop into existence. In general that could be a lot of things - like making entries here or there, making it known/available to whatever, checking and manipulating its context, setting up rules for it, attaching functions or members to it, registering event handlers to its events or whatnot.

Then there could be a lot of different EHs one might want to attach - onResize, onGainFocus, onReload, onClose... (don't know which actually are there).
In your case it's just an initial one-shot doing stuff within the new window. The question now is on what event to set it loose. The sensible thing of course is chosing an event that signals a certain state of the window's life cycle has been reached. So then one has to choose a state one wants to be reached.

To me it's just that the current choice is "the window merely exists" allowing the 2nd event handler (EH2) to be merged with EH1. But now one needs to check if the window's inner parts one wants to work on are already set up. Thus one might want to switch the state of choice to one guaranteeing those part's existence.
window.document's DOMContentLoaded should - to my understanding - be just the one, but has proven not to be. The next one would be window's "load" event and it looks to me, that it is.

> Given that the need for delaying is inevitable anyways, ...

Being event-driven is not the same as polling. ;)

One of the only two uncleanlynesses left would be having EH1 check if "load" is yet to be fired, thus the need to put the element attaching code in a function outside EH2, so that EH1 can execute it, instead.
But here I blame the API for not making sure all onOpenWindow handlers are done, before firing "load".
The other one comes from an uncomplete event system. You can't check if the window's "load" event already got fired. And I tried registering to window.document's "load" event, but I don't get one. On the other hand one can check window.document.readyState. If it's not "complete" or it is ("loading" or "interactive"), the window's "load" hasn't been fired yet. I don't know, if there's a gap between the document reaching "complete" and "load being fired, but it doesn't matter as the document being complete suffices - the window's inner parts needed will be there.
(BTW, DOMContentLoaded signals the change from "loading" to "interactive".)
